### PR TITLE
Serde allows external implementation of formats

### PR DESCRIPTION
**Pull request by Gilthoniel**
This changes serde to allow external format implementations and statically registered them according to which support one wants.

### CODE REVIEW

**Comment by Gilthoniel** [comment]
The document will be updated in a following PR.

File: mino/minogrpc/routing/routing.go [hunk 1]
```diff
@@ -79,82 +87,156 @@ func (t TreeRoutingFactory) GetAddressFactory() mino.AddressFactory {
...
-// VisitJSON implements serde.Factory. It deserializes the tree routing in JSON
-// format.
-func (t TreeRoutingFactory) VisitJSON(in serde.FactoryInput) (serde.Message, error) {
-	m := json.TreeRouting{}
-	err := in.Feed(&m)
-	if err != nil {
-		return nil, xerrors.Errorf("couldn't deserialize message: %v", err)
+// TreeRouting holds the routing tree of a network. It allows each node of the
+// tree to know which child it should contact in order to relay a message that
+// is in it sub-tree.
```
Comment (nkcr) [inline mino/minogrpc/routing/routing.go:129]: ...in itS sub-tree.

File: mino/minogrpc/routing/routing.go [hunk 2]
```diff
@@ -79,82 +87,156 @@ func (t TreeRoutingFactory) GetAddressFactory() mino.AddressFactory {
...
+
+		addrsBuf[i] = buf
+	}
 
-	sort.Stable(&addrsBuf)
+	if tmpl.rootIndex < 0 || tmpl.rootIndex >= len(addrs) {
+		return nil, xerrors.Errorf("invalid root index %d", tmpl.rootIndex)
+	}
+
+	// Extract the root before the shuffle. We don't do it so that it is
```
Comment (nkcr) [inline mino/minogrpc/routing/routing.go:214]: The second part of the comment seems wrong.

File: serde/json/mod.go [hunk 3]
```diff
@@ -1,86 +1,49 @@
...
 
-// FactoryInput is an implementation of the factory input.
+// JSONEngine is a context engine to marshal and unmarshal in JSON format.
 //
-// - implements serde.FactoryInput
-type factoryInput struct {
-	serde serde.Serializer
-	data  []byte
-}
+// - implements serde.FormatEngine
```
Comment (nkcr) [inline serde/json/mod.go:26]:
```suggestion
// - implements serde.ContextEngine
```

File: serde/json/mod.go [hunk 4]
```diff
@@ -1,86 +1,49 @@
...
-//
-// https://golang.org/pkg/encoding/json/
+// Package json implements the context engine for a the JSON format.
 package json
 
 import (
 	"encoding/json"
 
-	"go.dedis.ch/dela/internal/serdereflect"
+	// Static registration of the JSON formats. By having there here, it ensures
```
Comment (nkcr) [inline serde/json/mod.go:7]:
```suggestion
	// Static registration of the JSON formats. By having them here, it ensures
```

File: blockchain/skipchain/types/block.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+// GetIndex returns the index of the block since the genesis block.
+func (b SkipBlock) GetIndex() uint64 {
+	return b.Index
+}
+
+// GetGenesisID returns the genesis ID.
+func (b SkipBlock) GetGenesisID() []byte {
+	return append([]byte{}, b.GenesisID[:]...)
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply blockchain/skipchain/types/block.go:138]: The point is to make the return independent from the actual field. If you do that, you get the underlying array (from the slice) so you might change it.

File: dkg/pedersen/types/messages.go [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+		dhkey:     dhkey,
+		signature: sig,
+		nonce:     nonce,
+		cipher:    cipher,
+	}
+}
+
+// GetDHKey returns the Diffie-Helmann key in bytes.
+func (d EncryptedDeal) GetDHKey() []byte {
+	return append([]byte{}, d.dhkey...)
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply dkg/pedersen/types/messages.go:89]: Same as above, it returns an independent array.

File: blockchain/skipchain/types/block.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nkcr) [reply blockchain/skipchain/types/block.go:138]: From this example, it seems to return a copy: https://play.golang.org/p/To_n08wwKeZ

[the remark this answers is not in the document]
Comment (Gilthoniel) [reply blockchain/skipchain/types/block.go:138]: Oh right it will do a copy anyway. Nice, thanks !

**Pull request merged**